### PR TITLE
Set 'native-routing-cidr' from 'cluster-pool-ipv4-cidr'

### LINKED ISSUE #12186: Set 'native-routing-cidr' automatically from 'cluster-pool-ipv4-cidr' if ipam == cluster-pool
This option:
https://github.com/cilium/cilium/blob/0ee0458af776f770aa1e03a82419e96242ed3d7d/pkg/option/config.go#L1748-L1749

can automatically be set to the value passed to:
https://github.com/cilium/cilium/blob/37139c33fb791d375ca852a01e176b603810cd54/operator/option/config.go#L108-L110

1. Add a hidden flag to the daemon around this place:
https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573

2. Use that flag to here automatically based on the cluster-pool-ipv4-cidr here:
https://github.com/cilium/cilium/blob/0ee0458af776f770aa1e03a82419e96242ed3d7d/pkg/option/config.go#L2338-L2340

### PR DESCRIPTION
**Pull request by soumynathan**
This PR sets the 'native-routing-cidr' from 'cluster-pool-ipv4-cidr'
if the ipam==cluster-pool is set.

Fixes: #12186
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>

### CODE REVIEW

**Review by aanm — CHANGES_REQUESTED** [review]
@soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573

Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken

**Comment by soumynathan** [comment]
> @soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here
So in that case should we define a hidden DaemonConfig flag and we should still use the 'constant' defined fro 'IPAMOperatorV4CIDR' or can it hold any a random constant.
> 
> https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573
> 
> Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken

@aanm So do you think that as part of this PR we should move it (operatorOption.IPAMOperatorV4CIDR) out of the operator/option/config.go to 'pkg/option'.

**Comment by soumynathan** [comment]
> > @soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here
> > So in that case should we define a hidden DaemonConfig flag and we should still use the 'constant' defined fro 'IPAMOperatorV4CIDR' or can it hold any a random constant.
> > https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573
> > 
> > Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken
> 
> @aanm So do you think that as part of this PR we should move it (operatorOption.IPAMOperatorV4CIDR) out of the operator/option/config.go to 'pkg/option'.

Addressed review comments, but waiting for your response for moving the varible.

**Comment by tklauser** [comment]
> @soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here
> 
> https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573
> 
> Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken

Yes, I agree with @aanm that it should be moved to `pkg/option` so we can avoid depending on `pkg/option/operator` in the agent.

**Comment by tklauser** [comment]
> > @soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here
> > So in that case should we define a hidden DaemonConfig flag and we should still use the 'constant' defined fro 'IPAMOperatorV4CIDR' or can it hold any a random constant.
> > https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573
> > 
> > Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken
> 
> @aanm So do you think that as part of this PR we should move it (operatorOption.IPAMOperatorV4CIDR) out of the operator/option/config.go to 'pkg/option'.

Yes, please.

File: pkg/option/config.go
Comment (aanm) [inline pkg/option/config.go]: Interesting, I didn't know that `ipv4NativeRoutingCIDR` could only have 1 CIDR. This needs to be addressed first since we allow to have more than one CIDR in the `IPAMClusterPoolIPv4RoutingCIDR`. The problem rises with the fact that `native-routing-cidr` is a flag that can only receive a single CIDR and we can't change its behavior. Perhaps we would need to deprecate the `native-routing-cidr` and introduce a `native-routing-cidrs`. We should discuss this in the weekly meeting.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go]: @aanm, Yes, I had this question while I was working on it, since it was a array to a single value match. Ok, let us discuss it in the weekly meeting.

**Comment by soumynathan** [comment]
> > > @soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here
> > > So in that case should we define a hidden DaemonConfig flag and we should still use the 'constant' defined fro 'IPAMOperatorV4CIDR' or can it hold any a random constant.
> > > https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573
> > > 
> > > Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken
> > 
> > 
> > @aanm So do you think that as part of this PR we should move it (operatorOption.IPAMOperatorV4CIDR) out of the operator/option/config.go to 'pkg/option'.
> 
> Yes, please.

Sure will try to move it here in the /pkg/option once we have a clear idea of how we wanted to proceed with cidrs versus cidr.

**Comment by soumynathan** [comment]
> > > @soumynathan this won't be enough as the cilium-agent won't read the `operatorOption.IPAMOperatorV4CIDR` flag if it's not set here
> > > So in that case should we define a hidden DaemonConfig flag and we should still use the 'constant' defined fro 'IPAMOperatorV4CIDR' or can it hold any a random constant.
> > > https://github.com/cilium/cilium/blob/da6ed27b926701349f67b50524f977ad5a5e2ad0/daemon/cmd/daemon_main.go#L571-L573
> > > 
> > > Also, since the `operatorOption.IPAMOperatorV4CIDR` is not specific to the `operator` it needs to be moved out of that package into the `pkg/option`. cc @tklauser do double check if I'm not mistaken
> > 
> > 
> > @aanm So do you think that as part of this PR we should move it (operatorOption.IPAMOperatorV4CIDR) out of the operator/option/config.go to 'pkg/option'.
> 
> Yes, please.

Yes I have addressed your review comments to move the ClusterPool CIDR and Masks to the agent from the operator.

File: pkg/option/config.go
Comment (joestringer) [reply pkg/option/config.go]: I understand that we're looking for @brb 's input on this, he's currently OOO but will be back next week.

Comment (soumynathan) [reply pkg/option/config.go]: Ok, thanks will wait for the reply.

Comment (joestringer) [reply pkg/option/config.go]: @soumynathan were you able to get the feedback you need from @brb?

Comment (brb) [reply pkg/option/config.go]: We discussed during the meeting, that the current implementation of the BPF masq assumes that there will be an exactly one native routing CIDR. To overcome this limitation, we could reuse the existing BPF ip-masq-agent infrastructure, and to populate the CIDRs into its LPM BPF maps.

Comment (soumynathan) [reply pkg/option/config.go]: > We discussed during the meeting, that the current implementation of the BPF masq assumes that there will be an exactly one native routing CIDR. To overcome this limitation, we could reuse the existing BPF ip-masq-agent infrastructure, and to populate the CIDRs into its LPM BPF maps.

@brb @aanm, so do we need to modify this PR or drop it and work on a different one.

Comment (brb) [reply pkg/option/config.go]: I suggest to make a final decision during next week's community meeting.

Comment (soumynathan) [reply pkg/option/config.go]: @brb Thanks,

**Comment by nathanjsweet** [comment]
retest-4.19

**Comment by nathanjsweet** [comment]
retest-4.9

**Comment by nathanjsweet** [comment]
retest-net-next

File: pkg/option/config.go
Comment (aanm) [inline pkg/option/config.go]: We should do this regardless of IPAM being the operator or not. The user might have other CIDRs in mind to set in the `ipv4NativeRoutingCIDR` so we should not overwrite it as we are doing in 2360.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go]: Ok I will address your comments.

Comment (soumynathan) [reply pkg/option/config.go]: addressed review comments.

File: pkg/option/config.go
Comment (aanm) [inline pkg/option/config.go:2545]: Please add a comment here:
```golang
// the user might have other CIDRs not specified in the `IPAMOperatorV4CIDR` so we 
// should overwrite `ipv4NativeRoutingCIDR` with the values set in `IPv4NativeRoutingCIDR`
```

File: pkg/option/config.go
Comment (aanm) [inline pkg/option/config.go]: this can be removed

File: pkg/option/config.go
Comment (aanm) [inline pkg/option/config.go]: and we can initialize the variable here `if nativeCIDR := strings...`

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go]: will address your comments.

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go]: Done.

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go]: Done.

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go:2545]: Done

**Review by aanm — APPROVED** [review]
This needs a follow up to accept multiple pod CIDRs in the ipv4 native routing in the datapath. IPV4_SNAT_EXCLUSION_DST_CIDR

**Comment by aanm** [comment]
test-me-please

**Comment by brb** [comment]
> This needs a follow up to accept multiple pod CIDRs in the ipv4 native routing in the datapath. 

Can we create the issue before this PR gets merged?

File: pkg/option/config.go
Comment (brb) [inline pkg/option/config.go]: Could you explain what you are trying to achieve with this?

Comment (vadorovsky) [inline pkg/option/config.go]: `strings.Join(c.IPAMClusterPoolIPv4RoutingCIDR[:0], "")` will always return an empty string, so that statement is never going to be true.

`c.IPAMClusterPoolIPv4RoutingCIDR[:0]` will always return an empty slice, then joining it with an empty string doesn't do anything. See: https://play.golang.org/p/y7OWFpWTi-s

@soumynathan Can we maybe just use the first element directly?

```suggestion
	// TODO: Allow to set multiple CIDRs, use the entire slice.
	if len(c.IMAPClusterPoolIPv4RoutingCIDR) > 0 {
		c.ipv4NativeRoutingCIDR = cidr.MustParseCIDR(c.IMAPClusterPoolIPv4RoutingCIDR[0])
	}
```

**Review by vadorovsky — CHANGES_REQUESTED** [review]

File: pkg/option/config.go
Comment (soumynathan) [reply pkg/option/config.go]: Will address your comments.

**Comment by qmonnet** [comment]
test-me-please

**Review by vadorovsky — APPROVED** [review]
LGTM. Thanks!

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
retest-net-next

File: pkg/option/config.go
Comment (brb) [inline pkg/option/config.go:2544]: 1. Please log an info msg that native routing cidr is derived from the ipma cluster ipv4 pool.
2. What's the reasoning for using the first element from the slice? Perhaps you want to add a code comment explaining the reasoning.

**Review by brb — CHANGES_REQUESTED** [review]
Some questions/nits. Also, in the commit msg I'm missing explanation why the native routing cidr can be derived from cluster ipv4 routing cidr.

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
@soumynathan the tests are failing

File: pkg/option/config.go
Comment (vadorovsky) [reply pkg/option/config.go:2544]: ad 2. Actually I don't see the reason of `[]string{}` as a type of `IPAMClusterPoolIPv4RoutingCIDR`. I will change it just to `string`.

**Comment by mrostecki** [comment]
I will continue Swami's work on that and post a new PR with comments addressed (also with #12294 included, as both PRs address the same issue).

**Pull request closed without merging**
